File: src/protocol.ts
import { log } from './log'

/**
 * Bedrock point releases regularly bump the protocol *number* without
 * changing the protocol *schema* (1.26.40 -> 1.26.45 shipped as "Various bug
 * fixes" with zero packet changes), but the server hard-kicks any client
 * announcing an older number — `play_status: failed_client` before login
 * even starts. minecraft-data usually gains the new version entry days to
 * weeks later, and this image prunes all but one version's data anyway.
 *
 * So before each session: ping the server, and when it advertises a
 * different protocol number than the baked data maps to, re-map the baked
 * version to the advertised number. The session then announces the server's
 * own number while speaking the baked schema. On a schema-changing release
 * this trades a clean pre-login kick for a mid-join parse failure — same
 * reconnect loop either way, and the spoof log line names the culprit.
 */

const PING_TIMEOUT_MS = 5_000

interface Advertisement {
  protocol?: string | number
  version?: string
}

/**
 * Pure decision core: mutates `versions[version]` to match the advertised
 * protocol number. `baked` is what the library mapped `version` to before any
 * spoofing, so a server moving back (restore, rollback) clears the override.
 */
export function applyAdvertisedProtocol(
  versions: Record<string, number>,
  baked: number,
  version: string,
  advertisement: Advertisement,
): void {
  const advertised = Number(advertisement.protocol)
  if (!Number.isInteger(advertised) || advertised <= 0) {
    log('warn', 'protocol_advertisement_invalid', { protocol: advertisement.protocol ?? null })
    return
  }

  if (advertised === versions[version]) return

  if (advertised === baked) {
    versions[version] = baked
    log('info', 'protocol_spoof_cleared', { version, protocol: baked })
    return
  }

  versions[version] = advertised
  log('warn', 'protocol_spoofed', {
    version,
    baked_protocol: baked,
    advertised_protocol: advertised,
    server_version: advertisement.version ?? null,
  })
}

let bakedProtocols: Record<string, number> | undefined

/**
 * Ping the server and align the announced protocol number with what it
 * advertises. Never throws: on any ping failure the last known mapping
 * stands and the connection attempt proceeds — the server may be
 * mid-restart, which the caller's reconnect loop already handles.
 *
 * bedrock-protocol is required lazily so this module stays importable where
 * the native RakNet addon is not built (CI runs `npm ci --ignore-scripts`).
 */
export async function negotiateProtocol(host: string, port: number, version: string): Promise<void> {
  // Internal module path, not public API. bedrock-protocol is pinned to an
  // exact version in package.json, so the path and the shape of `Versions`
  // only move under a reviewed dependency bump.
  const { Versions } = require('bedrock-protocol/src/options') as { Versions: Record<string, number> }
  bakedProtocols ??= { ...Versions }
  const baked = bakedProtocols[version]
  if (baked === undefined) return

  const { ping } = require('bedrock-protocol') as {
    ping: (opts: { host: string; port: number }) => Promise<Advertisement>
  }

  let advertisement: Advertisement
  try {
    advertisement = await Promise.race([
      ping({ host, port }),
      new Promise<never>((_, reject) =>
        setTimeout(() => reject(new Error(`ping timed out after ${PING_TIMEOUT_MS}ms`)), PING_TIMEOUT_MS).unref(),
      ),
    ])
  } catch (error) {
    log('warn', 'protocol_ping_failed', { error: (error as Error).message })
    return
  }

  applyAdvertisedProtocol(Versions, baked, version, advertisement)
}

File: internal/mcproto/protocol.go
// Copied from github.com/jdwillmsen/minecraft-server-agent/pkg/mcproto.
//
// Copied rather than imported: importing would make the agent's dependency
// floors this bot's, and put every agent release up for review here, to share
// code upstream has stopped changing. docs/decisions.md has the full reasoning.
//
// Fix bugs upstream first, then port here.

// Package mcproto keeps a headless client connectable across Bedrock point
// releases that bump the protocol number without changing the packet schema.
//
// The server hard-kicks any client announcing an older protocol number --
// play_status: failed_client, before login even starts -- so a point release
// shipped as "various bug fixes" with no packet changes still takes every bot
// offline at once.
//
// So this pings the server before dialling and announces whatever number it
// advertises, while still speaking the schema this binary was built against.
// On a release that really does change the schema this trades a clean
// pre-login kick for a parse failure mid-join -- the same reconnect loop
// either way, and the substitution is logged by name so the cause is visible.
//
// docs/protocol-drift.md records the releases this has been needed for.
package mcproto

import (
	"context"
	"fmt"
	"strconv"
	"strings"
	"time"

	"github.com/sandertv/go-raknet"
	"github.com/sandertv/gophertunnel/minecraft"
)

// pingTimeout bounds the extra delay a ping adds to every connection attempt.
// A server that is mid-restart will not answer, and waiting on it is worse
// than dialling with the compiled-in number and letting the reconnect loop
// handle the failure.
const pingTimeout = 5 * time.Second

// Advertisement is what a server says about itself in its pong.
type Advertisement struct {
	Protocol int32
	Version  string
}

// Negotiate picks the Protocol to dial the server at address with.
//
// It never fails the caller: on any ping or parse problem it returns the
// compiled-in protocol, which is what a client without this package would
// have used. The optional onSpoof callback reports a decision worth logging;
// it is not called when the compiled-in number already matches.
func Negotiate(ctx context.Context, address string, onSpoof func(Advertisement)) minecraft.Protocol {
	ctx, cancel := context.WithTimeout(ctx, pingTimeout)
	defer cancel()

	pong, err := raknet.PingContext(ctx, address)
	if err != nil {
		return minecraft.DefaultProtocol
	}
	ad, err := ParsePong(string(pong))
	if err != nil {
		return minecraft.DefaultProtocol
	}
	return protocolFor(ad, minecraft.DefaultProtocol, onSpoof)
}

// protocolFor is the decision on its own, so it can be tested without a
// server to ping.
func protocolFor(ad Advertisement, baked minecraft.Protocol, onSpoof func(Advertisement)) minecraft.Protocol {
	if ad.Protocol == baked.ID() {
		return baked
	}
	if onSpoof != nil {
		onSpoof(ad)
	}
	return spoofed{Protocol: baked, id: ad.Protocol}
}

// ParsePong reads the protocol number and version out of a Bedrock pong.
//
// The pong is a semicolon-separated list whose leading fields have been
// stable for years, verified against the production server on 2026-09-08:
//
//	MCPE;FWB Server;2169;1.26.45;4;20;13467591190557326198;FWB;Survival;...
//	 0   1          2    3       4 5  6                    7   8
//
// Only fields 2 and 3 are read. Everything past them varies by server
// software and is none of this package's business.
func ParsePong(pong string) (Advertisement, error) {
	fields := splitPong(pong)
	if len(fields) < 4 {
		return Advertisement{}, fmt.Errorf("mcproto: pong has %d fields, want at least 4", len(fields))
	}
	n, err := strconv.ParseInt(strings.TrimSpace(fields[2]), 10, 32)
	if err != nil {
		return Advertisement{}, fmt.Errorf("mcproto: protocol field %q is not a number", fields[2])
	}
	// A server that advertises 0 or a negative number is broken or is not a
	// Bedrock server at all. Announcing that back would guarantee a kick,
	// where the compiled-in number at least has a chance.
	if n <= 0 {
		return Advertisement{}, fmt.Errorf("mcproto: protocol field is %d, want a positive number", n)
	}
	return Advertisement{Protocol: int32(n), Version: fields[3]}, nil
}

// splitPong splits on unescaped semicolons, matching how the server escapes
// them inside a MOTD.
func splitPong(s string) []string {
	var (
		fields []string
		cur    strings.Builder
		escape bool
	)
	for _, r := range s {
		switch {
		case escape:
			escape = false
			cur.WriteRune(r)
		case r == '\\':
			escape = true
		case r == ';':
			fields = append(fields, cur.String())
			cur.Reset()
		default:
			cur.WriteRune(r)
		}
	}
	return append(fields, cur.String())
}

// spoofed announces a different protocol number while delegating everything
// that touches packet layout to the protocol this binary was built against.
//
// Overriding ID alone is the whole point: change anything else and the client
// would be claiming to speak a schema it does not have.
type spoofed struct {
	minecraft.Protocol
	id int32
}

func (s spoofed) ID() int32 { return s.id }
